pkg/shp/cmd/follower/follow.go

```diff
@@ -0,0 +1,247 @@
+	}
+	f.logTail = tail.NewTail(f.ctx, clientset)
+	f.namespace = params.Namespace()
+	if f.shpClientset, err = params.ShipwrightClientSet(); err != nil {
+		return err
+	}
+	f.kclientset, err = params.ClientSet()
+	if err != nil {
+		return err
+	}
+	to, err := params.RequestTimeout()
+	if err != nil {
+		return err
+	}
+	f.pw, err = reactor.NewPodWatcher(f.ctx, to, f.kclientset, params.Namespace())
+	if err != nil {
+		return err
+	}
+
+	f.pw.WithOnPodModifiedFn(f.OnEvent)
+	f.pw.WithTimeoutPodFn(f.OnTimeout)
+	f.pw.WithNoPodEventsYetFn(f.OnNoPodEventsYet)
+	return nil
+}
+
+// Log prints a message
+func (f *Follower) Log(msg string) {
+	// concurrent fmt.Fprintf(r.ioStream.Out...) calls need locking to avoid data races, as we 'write' to the stream
+	f.logLock.Lock()
+	defer f.logLock.Unlock()
+	fmt.Fprint(f.ioStreams.Out, msg)
+}
+
+// tailLogs start tailing logs for each container name in init-containers and containers, if not
+// started already.
+func (f *Follower) tailLogs(pod *corev1.Pod) {
+	containers := append(pod.Spec.InitContainers, pod.Spec.Containers...)
+	for _, container := range containers {
+		if _, exists := f.tailLogsStarted[container.Name]; exists {
+			continue
+		}
+		f.tailLogsStarted[container.Name] = true
+		f.logTail.Start(pod.GetNamespace(), pod.GetName(), container.Name)
+	}
+}
+
+// stop invoke stop on streaming components.
+func (f *Follower) stop() {
+	f.logTail.Stop()
+	f.pw.Stop()
+}
+
+// OnEvent reacts on pod state changes, to start and stop tailing container logs.
+func (f *Follower) OnEvent(pod *corev1.Pod) error {
+	switch pod.Status.Phase {
+	case corev1.PodRunning:
+		if !f.enteredRunningState {
+			f.Log(fmt.Sprintf("Pod %q in %q state, starting up log tail", pod.GetName(), corev1.PodRunning))
+			f.enteredRunningState = true
+			// graceful time to wait for container start
+			time.Sleep(3 * time.Second)
+			// start tailing container logs
+			f.tailLogs(pod)
+		}
+	case corev1.PodFailed:
+		msg := ""
+		var br *buildv1alpha1.BuildRun
+		err := wait.PollImmediate(1*time.Second, 15*time.Second, func() (done bool, err error) {
+			br, err = f.shpClientset.ShipwrightV1alpha1().BuildRuns(pod.Namespace).Get(f.ctx, f.buildRunName, metav1.GetOptions{})
+			if err != nil {
+				if kerrors.IsNotFound(err) {
+					return true, nil
+				}
+				f.Log(fmt.Sprintf("error getting buildrun %q for pod %q: %s\n", f.buildRunName, pod.GetName(), err.Error()))
+				return false, nil
+			}
+			if br.IsDone() {
+				return true, nil
+			}
+			return false, nil
+		})
+		if err != nil {
+			f.Log(fmt.Sprintf("gave up trying to get a buildrun %q in a terminal state for pod %q, proceeding with pod failure processing", f.buildRunName, pod.GetName()))
+		}
+		switch {
+		case err == nil && br.IsCanceled():
+			msg = fmt.Sprintf("BuildRun %q has been canceled.\n", br.Name)
+		case (err == nil && br.DeletionTimestamp != nil) || (err != nil && kerrors.IsNotFound(err)):
+			msg = fmt.Sprintf("BuildRun %q has been deleted.\n", br.Name)
```

Comment:
With your changes, `br` can be `nil` here when the condition in line 136 was met in the first iteration.

```suggestion
		case br == nil:
			msg = fmt.Sprintf("BuildRun %q has been deleted.\n", br.Name)
		case err == nil && br.IsCanceled():
			msg = fmt.Sprintf("BuildRun %q has been canceled.\n", br.Name)
		case (err == nil && br.DeletionTimestamp != nil) || (err != nil && kerrors.IsNotFound(err)):
			msg = fmt.Sprintf("BuildRun %q has been deleted.\n", br.Name)
```

Reply:
I've seen various k8s versions and forks vary a bit with returning nil vs. empty struct, but regardless, better safe than sorry - thx